### DISCUSSION TITLE
Is there any way to represent a composite value in filter?

### DISCUSSION THREAD

**Discussion by ondrowan** [diesel-rs/diesel#3658] [Q&A] [answered]
Postgres has a support for composite values in `WHERE` clause, e.g.:

```sql
SELECT *
FROM table
WHERE (datetime, public_id) < ('2023-05-29T08:41:55.610Z', 'f5a59c63-d8ca-401d-835a-2266bd7981ea')
ORDER BY datetime DESC, public_id DESC
LIMIT 10
```

Using a tuple in `.filter` doesn't seem to work since it complains about values in given tuple not implementing `PartialOrd` and `PartialEq` and I'm not sure how to fix it.

There are 2 ways how I was able to represent this.

As a longer representation of this kind of query:

```rust
query = query.filter(table::datetime.lt(datetime).or(
    (table::datetime.eq(datetime)).and(table::public_id.lt(public_id)),
));
```

Or manually:

```rust
query = query.filter(
    sql::<sql_types::Bool>("(datetime, public_id) < ").bind::<sql_types::Record<(
        sql_types::Timestamptz,
        sql_types::Uuid,
    )>, _>((
        datetime,
        public_id,
    ))
);
```

Is there a simpler way to write this kind of query?

**Comment by weiznich** [comment, marked as answer]
Such expressions are currently not supported by the built-in DSL. You can provide your own DSL extension for that by implementing `QueryFragment`, `QueryId` and `Expression` for a new type wrapper around a tuple.
